src/faz/server/faz_server.cpp

```diff
@@ -0,0 +1,65 @@
+#include "faz_server.h"
+
+FazServiceImpl::FazServiceImpl() {
+  // Initializing the kvstore client to connect
+  // with its backend over grpc
+  std::string target_str = "localhost:50001";
+  kvstore_(grpc::CreateChannel(target_str, grpc::InsecureChannelCredentials()));
+  // Initializing caw_functions_ to store the function
+  // data
+  caw_functions_["registeruser"] = RegisterUser;
+  caw_functions_["caw"] = PostCaw;
+  caw_functions_["read"] = ReadCaw;
+  caw_functions_["follow"] = FollowUser;
+  caw_functions_["profile"] = GetProfile;
+}
+
+Status FazServiceImpl::hook(ServerContext *context, const HookRequest *request,
+                            HookReply *response) {
+  int eventType = request->event_type();
+  std::string functionName = request->event_function();
+  Status status;
+  // If the key does not exist
+  if (caw_functions_.find(functionName) == caw_functions_.end()) {
+    status = Status(StatusCode::NOT_FOUND, "The given function does not exist");
+  } else {
+    hooked_functions_[eventType] = functionName;
+    status = Status::OK;
+  }
+  return status;
+}
+
+Status FazServiceImpl::unhook(ServerContext *context,
+                              const UnhookRequest *request,
+                              UnhookReply *response) {
+  int eventType = request->event_type();
+  Status status;
+  // If the key does not exist
+  if (hooked_functions_.find(eventType) == hooked_functions_.end()) {
+    status =
+        Status(StatusCode::NOT_FOUND, "The given event type does not exist");
+  } else {
+    hooked_functions_.erase(eventType);
+    status = Status::OK;
+  }
+  return status;
+}
+
+Status FazServiceImpl::event(ServerContext *context,
+                             const EventRequest *request,
+                             EventReply *response) {
+  int eventType = request->event_type();
+  Status status;
+  if (hooked_functions_.find(eventType) == hooked_functions_.end()) {
+    status =
+        Status(StatusCode::NOT_FOUND, "The given event type does not exist");
+  } else {
+    std::string functionName = hooked_functions_[eventType];
+    // Now we will call the function using the stored
+    // caw functions. Functions will modify data inside
+    // the Any objects.
+    status = caw_functions_[functionName](request->payload(),
+                                          response->payload(), kvstore_);
```

Comment:
Doesn't this access a const reference of payload?  If so, the caw function can't modify it.

Reply:
Great catch! That thought completely escaped me. The idea is that the caw function can just edit the objects and I wont have to create any weird looking structs to return data and move it through grpc. Would you recommend using mutable_payload() instead? I can pass a pointer to the object in and then modify it from there?

Reply:
You could also modify the return type of caw functions to be an optional<Any> to wrap the payload/success status. But, if you want to keep the return type as a status, then you could use the mutable payload approach.

Reply:
I decided to go with the mutable option. I think the code is much cleaner without having to return and parse wrapper structs. Have made the change!

Reply:
Ok so I am using request->payload() for the request because we do not need to change that. 
For the client i am using mutable_payload() but because that returns a pointer I dereference the pointer and pass in the object by reference to the client functions.